x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/StSimplify.java

```diff
@@ -0,0 +1,164 @@
+        ) Expression tolerance
+    ) {
+        super(source, List.of(geometry, tolerance));
+        this.geometry = geometry;
+        this.tolerance = tolerance;
+    }
+
+    private StSimplify(StreamInput in) throws IOException {
+        this(Source.readFrom((PlanStreamInput) in), in.readNamedWriteable(Expression.class), in.readNamedWriteable(Expression.class));
+    }
+
+    @Override
+    public DataType dataType() {
+        return tolerance.dataType() == DataType.NULL ? DataType.NULL : geometry.dataType();
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        return new StSimplify(source(), newChildren.get(0), newChildren.get(1));
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, StSimplify::new, geometry, tolerance);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        source().writeTo(out);
+        out.writeNamedWriteable(geometry);
+        out.writeNamedWriteable(tolerance);
+    }
+
+    private static BytesRef geoSourceAndConstantTolerance(BytesRef inputGeometry, double inputTolerance) {
+        if (inputGeometry == null) {
+            return null;
+        }
+        try {
+            org.locationtech.jts.geom.Geometry jtsGeometry = UNSPECIFIED.wkbToJtsGeometry(inputGeometry);
+            org.locationtech.jts.geom.Geometry simplifiedGeometry = DouglasPeuckerSimplifier.simplify(jtsGeometry, inputTolerance);
+            return UNSPECIFIED.jtsGeometryToWkb(simplifiedGeometry);
+        } catch (ParseException e) {
+            throw new IllegalArgumentException("could not parse the geometry expression: " + e);
+        }
+    }
+
+    @Override
+    public EvalOperator.ExpressionEvaluator.Factory toEvaluator(ToEvaluator toEvaluator) {
+        EvalOperator.ExpressionEvaluator.Factory geometryEvaluator = toEvaluator.apply(geometry);
+
+        if (tolerance.foldable() == false) {
+            throw new IllegalArgumentException("tolerance must be foldable");
+        }
+        var toleranceExpression = tolerance.fold(toEvaluator.foldCtx());
+        double inputTolerance = getInputTolerance(toleranceExpression);
+        return new StSimplifyNonFoldableGeometryAndFoldableToleranceEvaluator.Factory(source(), geometryEvaluator, inputTolerance);
+    }
+
+    private static double getInputTolerance(Object toleranceExpression) {
+        double inputTolerance;
+
+        if (toleranceExpression instanceof Integer) {
+            inputTolerance = ((Integer) toleranceExpression).doubleValue();
+        } else if (toleranceExpression instanceof Double) {
+            inputTolerance = (double) toleranceExpression;
+        } else {
+            throw new IllegalArgumentException("tolerance for st_simplify must be a non negative integer or double");
```

Comment:
I think we should support all `Number` instances. Even if the literal tolerance will always be one of these two types, if we read the tolerance from a field, or calculate it with an expression, we might get other valid types.